### ISSUE TITLE
Inheritance and type narrowing prevent private member access in base class.

### ISSUE THREAD

**Issue by panthus** [microsoft/TypeScript#11454]
**TypeScript Version:**  2.0.3

**Code**

``` ts
class Animal {
    private sound: string;

    public makeSound() {
        if (this instanceof Cat || this instanceof Dog) {
            this.sound = "grr"; // Errors
        }
    }
}

class Cat extends Animal {
    public walk() {}
}

class Dog extends Animal {
    public eat() {}
}
```

**Expected behavior:**
Should be able to access private member in its declaring class.

**Actual behavior:** 
Cannot access private member in its declaring class if the subclasses have different members and we narrow the type of `this` to a union type of its subclasses.

**Comment by falsandtru** [comment]
I think `this` narrowed by subclasses should have external accessibility of them. So I think this is a correct behavior.

**Comment by wallverb** [comment]
I think it should work. Works in Java. https://goo.gl/Pb5p2Y

**Comment by mhegazy** [comment]
So we should narrow to `Cat & Animal | Dog & Animal` ?

**Comment by jakebailey** [comment]
Just looking at old bugs; this one was fixed in 3.9+ via #38277, though I'm not sure that the PR actually introduces a test for this particular case (though the underlying thing is in fact the same; `this` is narrowed to `(this & Cat) | (this & Dog)`).
